README.md
![GitHub issues](https://img.shields.io/github/issues/CICCIOSGAMINO/progress-ring)
[![npm version](https://badgen.net/npm/v/@cicciosgamino/progress-ring)](https://www.npmjs.com/package/@cicciosgamino/progress-ring)
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/cicciosgamino/progress-ring)

# 🪐 \<progress-ring\>

Simple customizable progress ring 🪐!

### Example One

![Example One](https://raw.githubusercontent.com/CICCIOSGAMINO/cicciosgamino.github.io/master/images/exampleOne.gif)

```css
progress-ring {
  --progress-from-bottom: 10%;
  --progress-from-left: 50%;
  --bk-color: #34495e;
  --progress-stroke-color: #3598dc;
  --progress-opacity: 1;
  --progress-font-family: 'Roboto', sans-serif;
  --progress-font-size: 3rem;
  --progress-font-color: whitesmoke;
}
```

```html
<progress-ring stroke="20" radius="80" progress="0"></progress-ring>
```

### Example Two

![Example Two](https://raw.githubusercontent.com/CICCIOSGAMINO/cicciosgamino.github.io/master/images/exampleTwo.gif)

```css
progress-ring {
  --progress-from-bottom: 10%;
  --progress-from-left: 50%;
  --bk-color: transparent;
  --footprint-stroke-color: grey;
  --footprint-opacity: .1;
  --progress-stroke-color: #e9008d;
  --progress-stroke-gradient-color: #dc0030ff;
  --progress-stroke-linecap: round;
  --progress-font-family: 'Jua', sans-serif;
  --progress-font-size: 3rem;
  --progress-font-color: #dc0030ff;
  --progress-font-weight: 600;
}
```

```html
<progress-ring stroke="13" radius="80" progress="10" ></progress-ring>
```

### Example Three

![Example Four](https://raw.githubusercontent.com/CICCIOSGAMINO/cicciosgamino.github.io/master/images/exampleThree.gif)

```css
progress-ring {
  --progress-from-bottom: 5%;
  --progress-from-left: 50%;

  --bk-color: transparent;

  --progress-stroke-color: purple;
  --progress-stroke-gradient-color: #00bc9b;
  --progress-stroke-linecap: round;
  --progress-opacity: .9;

  --footprint-stroke-color: grey;
  --footprint-opacity: .2;

  --progress-font-family: 'Mitr', sans-serif;
  --progress-font-size: 3rem;
  --progress-font-weight: 700;
  --progress-font-color: #232323;
}
```

```html
<progress-ring stroke="13" radius="80" progress="0"></progress-ring>
```


### Example Four

![Example Four](https://raw.githubusercontent.com/CICCIOSGAMINO/cicciosgamino.github.io/master/images/exampleFour.gif)

```css
progress-ring {
  --progress-from-bottom: 10%;
  --progress-from-left: 50%;
  --title-from-top: 10%;
  --title-from-left: 50%;

  --bk-color: transparent;

  --progress-stroke-color: OrangeRed ;
  --progress-stroke-gradient-color: salmon;
  --progress-stroke-linecap: round;
  --progress-opacity: .9;

  --footprint-stroke-color: grey;
  --footprint-opacity: .2;

  --progress-font-family: 'Roboto', sans-serif;
  --progress-font-size: 2rem;
  --progress-font-weight: 700;
  --progress-font-color: salmon;

  --title-font-size: 4rem;
}
```

```html
<progress-ring title="🍣" stroke="10" radius="80" progress="0"></progress-ring>
```

## 🚀 Usage

1. Install package
```bash
npm install --save @cicciosgamino/progress-ring
```

2. Import
```html
<!-- Import Js Module -->
<script type="module">
  // Importing this module registers <progress-ring> as an element that you
  // can use in this page.
  //
  // Note this import is a bare module specifier, so it must be converted
  // to a path using a server such as @web/dev-server.
  import '@cicciosgamino/progress-ring'
</script>
```

3. Place in your HTML
```html
<progress-ring title="🍣" stroke="10" radius="80" progress="0"></progress-ring>
```

## 🐝 API

### 📒 Properties/Attributes

| Name | Type | Default | Description
| ------------- | ------------- | ------------ | --------------
| title    | String | `''` | Brief string to place inside the progress ring component space
| radius   | Number |  60  | Radius of the ring
| progress | Number |  0   | Percent number passed to the component
| stroke   | Number |  10  | Stroke width, this rapresent the progress line

### Methods
*None*

### Events
*None*

### 🧁 CSS Custom Properties

| Name | Default | Description
| -------------------------- | ----------- | --------------------
| `--progress-from-bottom`   |    `10%`    | Absolute positioning the progress number from bottom
| `--progress-from-left`     |    `50%`    | Absolute positioning the progress number from left 
| `--title-from-top`         |    `40%`    | Absolute positioning the title from top
| `--title-from-left`        |    `50%`    | Absolute positioning the title from left
| `--bk-color`               |`transparent`| Background color inside the circle
| `--progress-stroke-color`  |  `#e9008d`  | Progress stroke color
| `--progress-stroke-gradient-color`  |    `''`    | Progress stroke stop-color if you want a gradient effect
| `--progress-stroke-linecap`|    `butt`   | SVG stroke linecap ( butt | round | square )
| `--progress-opacity`       |     `.9`    | Opacity on progress stroke
| `--footprint-stroke-color` |`transparent`| Track under the progress stroke can be visible or transparent
| `--footprint-opacity     ` |     `0`     | Opacity on footprint stroke
| `--progress-font-family`   |`sans-serif` | Font family for the progress counter
| `--progress-font-size`     |    `2rem`   | Font size for the progress counter
| `--progress-font-weight`   |    `400`    | Font weight for the progress counter
| `--progress-font-color`    |    `#000`   | Font color for the progress counter
| `--title-font-family`      |`sans-serif` | Font family for the title
| `--title-font-size`        |   `1.2rem`  | Font size for the title
| `--title-font-color`       |    `#000`   | Font color for the title



### 🤖 Write HTML and JavaScript
Import the component's JavaScript module, use the component in your HTML, and control it with JavaScript, just like you would with a built-in element such as `<button>`:

```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>My Example App</title>

    <!-- Add support for Web Components to older browsers. -->
    <script src="./node_modules/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>

  </head>
  <body>
    <!-- Use Web Components in your HTML like regular built-in elements. -->
    <progress-ring id="myProgress" title="🍣" stroke="10" radius="80" progress="0"></progress-ring>

    <!-- The Material Web Components use standard JavaScript modules. -->
    <script type="module">

      // Importing this module registers <progress-ring> as an element that you
      // can use in this page.
      //
      // Note this import is a bare module specifier, so it must be converted
      // to a path using a server such as @web/dev-server.
      import '@cicciosgamino/progress-ring'

      // Standard DOM APIs work with Web Components just like they do for
      // built-in elements.
      const p = document.querySelector('progress-ring')
      let count = 0
      setInterval(() => {
        p.setAttribute('progress', `${(count+=1) % 100}`)
      }, 100)
    </script>
  </body>
</html>
```

### 🚀 Serve
Serve your HTML with any server or build process that supports *bare module specifier resolution* (see next section):

```sh
# use globally instelled
npm install -g @web/dev-server

# install the project dev-dependencies and npm run
npm install
npm run serve
```

## Bare module specifiers

All @cicciosgamino Lit components [JavaScript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) that use *bare module specifiers*. Bare module specifiers are not yet supported by browsers, so it is necessary to use a tool that transforms
them to a *path* (for example from `@cicciosgamino/progress-ring` to `./node_modules/@cicciosgamino/progress-ring/progress-ring.js`).

Two great choices for tools that do this are:

- During local development, use open-wc's [`@web/dev-server`](https://www.npmjs.com/package/@web/dev-server) with the `--node-resolve` flag.
- For your production deployment, build your application with [Rollup](https://rollupjs.org/guide/en/) using the [`rollup-plugin-node-resolve`](https://github.com/rollup/rollup-plugin-node-resolve) plugin.

## Contributing

Got **something interesting** you'd like to **share**? Learn about [contributing](https://github.com/CICCIOSGAMINO/init/blob/master/CONTRIBUTING.md).

## Author

| [![@cicciosgamino](https://raw.githubusercontent.com/CICCIOSGAMINO/cicciosgamino.github.io/master/images/justme%40412x412_round.png)](https://linkedin.com/in/) 	|
|:------------------------------------------------------------------------------------------:	|
|                                    **@cicciosgamino**                                      	|

## Support
Reach out to me at one of the following places:

- [Github](https://github.com/CICCIOSGAMINO)
- [Twitter](https://twitter.com/cicciosgamino)

## Donate

Donate help and contibutions!

## License

![GitHub](https://img.shields.io/github/license/cicciosgamino/progress-ring)

- Copyright © [@cicciosgamino](https://cicciosgamino.web.app).


## CSS encapsulation
The elements HTML structure is initialized in a [Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM), so it is impossible to apply CSS to it. If you need to change the element's default style for any reason, you should open up a new issue (or a pull request!), describing your use case, and we'll work with you on solving the problem.

## Browser support

Browsers without native [custom element support][support] require a [polyfill][]. Legacy browsers require various other polyfills. See [`examples/index.html`][example] for details.

- Chrome
- Firefox
- Safari
- Microsoft Edge

[support]: https://caniuse.com/#feat=custom-elementsv1
[polyfill]: https://github.com/webcomponents/custom-elements

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 18-05-2021
### Added
  - Add default transparent to the circle background

### Changed
  - Minor into README.md

## [1.0.0] - 18-05-2021
### Added
  - Init the Element  [@cicciosgamino](https://github.com/CICCIOSGAMINO).

progress-ring.js
import { LitElement, html, css } from 'lit'

class ProgressRing extends LitElement {
  #normalizedRadius
  #circumference
  #strokeDasharray
  #strokeDashoffset
  static get styles () {
    return css`
      :host {
        /* positioning the progress and the title
        --progress-from-bottom: 10%;
        --progress-from-left: 50%;
        --title-from-top: 40%;
        --title-from-left: 50%;
        /* background color of the circle 
        --bk-color: transparent; */
        /* progress indicator color / stop gradient-color / opacity 
        --progress-stroke-color: purple;
        --progress-stroke-gradient-color: #00bc9b;
        --progress-stroke-linecap: round;
        --progress-opacity: .9; */
        /* trace of the indicator path and opacity
        --footprint-stroke-color: grey;
        --footprint-opacity: .2; */
        /* progress text style 
        --progress-font-family: 'Roboto', sans-serif;
        --progress-font-size: 2rem;
        --progress-font-weight: 700;
        --progress-font-color: salmon; */
        /* title text style 
        --title-font-family: 'Roboto', sans-serif;
        --title-font-size: 1.3rem;
        --title-font-color: blue; */

        display: inline-block;
      }

      .content {
        position: relative;
        text-align: center;
      }

      .foodprint-ring__circle {
        opacity: var(--footprint-opacity, 0);
        stroke: var(--footprint-stroke-color, transparent);
      }

      .progress-ring__circle {
        fill: var(--bk-color);
        stroke-linecap: var(--progress-stroke-linecap, butt);
        /* stroke: var(--progress-stroke-color); */
        opacity: var(--progress-opacity, .9);
        /* One particular thing about stroke-dashoffset, its starting point 
          is vertically centered and horizontally tilted to the right. It’s 
          necessary to negatively rotate the circle to get the desired effect. */
        transform: rotate(-90deg);
        transform-origin: 50% 50%;
        /* by transitioning the property we will get the animation feel */
        transition: stroke-dashoffset 0.35s;
      }

      #gradient {
        color: var(--progress-stroke-color, #e9008d);
      }

      #stop {
        color: var(--progress-stroke-gradient-color, var(--progress-stroke-color));
      }

      .title {
        font-family: var(--title-font-family, sans-serif);
        font-size: var(--title-font-size, 1.2rem);
        color: var(--title-font-color, #000);
        position: absolute;
        top: var(--title-from-top, 40%);
        left: var(--title-from-left, 50%);
        transform: translate(-50%, -50%);
      }

      .progress {
        font-family: var(--progress-font-family, sans-serif);
        font-size: var(--progress-font-size, 2rem);
        font-weight: var(--progress-font-weight, 400);
        color: var(--progress-font-color, #000);
        position: absolute;
        bottom: var(--progress-from-bottom, 10%);
        left: var(--progress-from-left, 50%);
        transform: translate(-45%, -45%);
      }
    `
  }

  static get properties () {
    return {
      title: String,
      radius: Number,
      progress: Number,
      stroke: Number
    }
  }

  constructor () {
    super()
    this.title = ''
    this.radius = 60
    this.progress = 0
    this.stroke = 10
  }

  connectedCallback () {
    super.connectedCallback()
    this.#normalizedRadius = this.radius - this.stroke
    this.#circumference = this.#normalizedRadius * 2 * Math.PI
    this.#strokeDasharray = `${this.#circumference} ${this.#circumference}`
    this.#strokeDashoffset = 
        this.#circumference - this.progress / 100 * this.#circumference
  }

  update (changed) {
    // handle the progress updated
    if (changed.has('progress')) {
      this.#strokeDashoffset = 
        this.#circumference - this.progress / 100 * this.#circumference
    }
    super.update()
  }

  render () {
    return html`
    <div class="content">

      <svg class="progress-ring" 
        height=${this.radius * 2} 
        width=${this.radius * 2} >

        <defs>
          <linearGradient id="gradient" x1="0%" y1="0%" x2="0%" y2="100%">
            <stop offset="0%"  stop-color="currentcolor" />
            <stop id="stop" offset="100%" stop-color="currentcolor" />
          </linearGradient>
        </defs>

        <circle 
          class="foodprint-ring__circle"
          shape-rendering="geometricPrecision"
          stroke-width=${this.stroke}
          fill="transparent"
          r=${this.#normalizedRadius}
          cx=${this.radius}
          cy=${this.radius} />

        <circle 
          class="progress-ring__circle"
          shape-rendering="geometricPrecision"
          stroke="url(#gradient)"
          stroke-width=${this.stroke}
          r=${this.#normalizedRadius}
          cx=${this.radius}
          cy=${this.radius}
          stroke-dasharray=${this.#strokeDasharray}
          stroke-dashoffset=${this.#strokeDashoffset} />

      </svg>

      <p class="title">${this.title}</p>
      <p class="progress">${this.progress}%</p>

  </div>
    `
  }
}

customElements.define('progress-ring', ProgressRing)
